xla/pjrt/gpu/se_gpu_pjrt_client.cc

```diff
@@ -406,140 +660,208 @@ StreamExecutorGpuClient::CrossHostSendBuffers(
-                          kv_store_->Get(cross_host_transfer_name,
-                                         cross_host_transfer_timeout_));
-      CliqueId clique_id(descriptor);
-
-      // Create a communicator.
-      TF_ASSIGN_OR_RETURN(
-          std::unique_ptr<Communicator> communicator,
-          CreateTransferCommunicator(local_device, gpu_collectives, clique_id,
-                                     /*is_sender=*/true));
-
-      // Send data to the receiver.
+      // Launch the send.
       auto mem = tensorflow::down_cast<PjRtStreamExecutorRawBuffer*>(
-                     held_send_buffer.raw_buffer.get())
+                     prepared_send.raw_buffer.get())
                      ->device_buffer();
+      TF_RETURN_IF_ERROR(gpu_communicator->LaunchSend(
+          /*send_buffer=*/mem->mem(),
+          /*dtype=*/xla::PrimitiveType::U8,
+          /*count=*/mem->mem().size(),
+          /*peer=*/RankId(1),
+          /*executor=*/gpu::GpuCollectives::On(*stream)));
+    }
+    return absl::OkStatus();
+  };
 
-      Future<> send_future = communicator->Send(
-          mem->mem(), shape.element_type(), ShapeUtil::ElementsIn(shape),
-          RankId(0), gpu::GpuCollectives::On(*stream));
-      TF_RETURN_IF_ERROR(send_future.Await());
-
-      // Mark send as done.
-      TF_RETURN_IF_ERROR(
-          AllocateAndRecordEvent(usage_event->event(), local_device, stream));
-
-      return absl::OkStatus();
-    };
-
-    absl::Status s = f();
-    if (!s.ok()) {
-      SetEventAsError(usage_event->event(), s);
+  // Form the closure to schedule on the device's execute thread.
+  auto execute_sends_fn = [this, local_device_state, stream,
+                           promises = std::move(promises),
+                           prepared_sends = std::move(prepared_sends),
+                           launch_send_group = std::move(launch_send_group),
+                           usage_event = std::move(usage_event)]() mutable {
+    // Group transfers by GPU clique.
+    absl::flat_hash_map<gpu::GpuCliqueKey, std::vector<PreparedSend>>
+        grouped_sends = GroupSendsByCliqueKey(std::move(prepared_sends));
+
+    // Transfers for a particular clique are executed as a group. This vector
+    // holds group futures for each clique_key in grouped_sends.
+    std::vector<Future<>> group_futures;
+    group_futures.reserve(grouped_sends.size());
+
+    for (auto& [clique_key, curr_sends] : grouped_sends) {
+      // Get the communicator on which we will execute this group of
+      // transfers. We assume each clique key is associated with a unique
+      // communicator, so we just take the communicator of the first
+      // transfer_idx of this clique key.
+      gpu::GpuCommunicator* gpu_communicator =
+          curr_sends[0].clique_and_communicator.second;
+
+      // Launch the group of transfers.
+      group_futures.push_back(gpu_communicator->GroupExecute(
+          [&launch_send_group, &curr_sends,
```

Comment:
Linter says "captured structured bindings are incompatible with C++ standards before C++20" for `&curr_sends`. I think it can be fixed by using `&curr_sends = curr_sends` in the capture instead. (Same with `curr_receives` on 1029).

Sorry for adding these linter comments piecemeal, I think that's the last of it.